examples/language-modeling/run_electra_pretraining.py

```diff
@@ -0,0 +1,752 @@
+
+    @staticmethod
+    def gather_positions(sequence, positions):
+        batch_size, sequence_length, dimension = sequence.shape
+        position_shift = (sequence_length * torch.arange(batch_size, device=sequence.device)).unsqueeze(-1)
+        flat_positions = torch.reshape(positions + position_shift, [-1]).long()
+        flat_sequence = torch.reshape(sequence, [batch_size * sequence_length, dimension])
+        gathered = flat_sequence.index_select(0, flat_positions)
+        return torch.reshape(gathered, [batch_size, -1, dimension])
+
+    def forward(
+        self, input_ids=None, attention_mask=None, token_type_ids=None, position_ids=None, head_mask=None, labels=None,
+    ):
+        # get the masked positions as well as their original values
+        masked_input_ids, masked_lm_positions = self.mask_inputs(
+            input_ids, [self.tokenizer.cls_token_id, self.tokenizer.sep_token_id, self.tokenizer.mask_token_id],
+        )
+
+        # only masked values should be counted in the loss; build a tensor containing the true values and -100 otherwise
+        masked_lm_labels = torch.full_like(input_ids, -100)
+        masked_lm_labels.scatter_(-1, masked_lm_positions, masked_input_ids)
+
+        # Create a tensor filled with masks
+        masked_tokens = torch.full_like(masked_input_ids, self.tokenizer.mask_token_id)
+        masked_lm_inputs = input_ids.clone()
+
+        # Of the evaluated tokens, 15% of those will keep their original tokens
+        replace_with_mask_positions = masked_lm_positions * (
+            torch.rand(masked_lm_positions.shape, device=masked_lm_positions.device) < (1 - self.mask_probability)
+        )
+
+        # Scatter the masks at the masked positions
+        masked_lm_inputs.scatter_(-1, replace_with_mask_positions, masked_tokens)
+        masked_lm_inputs[..., 0] = 101
+
+        generator_loss, generator_output = self.generator(
+            masked_lm_inputs,
+            attention_mask,
+            token_type_ids,
+            position_ids,
+            head_mask,
+            position_ids,
+            masked_lm_labels=masked_lm_labels,
+        )[:2]
+
+        # get the generator's predicted value on each masked position
+        fake_logits = self.gather_positions(generator_output, masked_lm_positions)
+        fake_argmaxes = fake_logits.argmax(-1)
+
+        # create a tensor containing the predicted tokens
+        fake_tokens = input_ids.scatter(-1, masked_lm_positions, fake_argmaxes)
+        fake_tokens[:, 0] = input_ids[:, 0]
+        discriminator_labels = (labels != fake_tokens).int()
+
+        discriminator_loss, discriminator_output = self.discriminator(
+            fake_tokens,
+            attention_mask,
+            token_type_ids,
+            position_ids,
+            head_mask,
+            position_ids,
+            labels=discriminator_labels,
+        )[:2]
+
+        discriminator_predictions = torch.round((torch.sign(discriminator_output) + 1) / 2)
```

Comment:
Maybe use numbers instead of integers for `+ 1.0` and `/ 2.0` to avoid type dispatching. 
Also, multiplication is faster than division `* 0.5`, but will have negligeable impact.